<a href="https://dewey.tailorbrands.com/production/brand_version_mockup_image/728/3420799728_89d22a7a-16da-4584-b076-06933635f74e.png?cb=1596001276"><img src = "https://dewey.tailorbrands.com/production/brand_version_mockup_image/728/3420799728_89d22a7a-16da-4584-b076-06933635f74e.png?cb=1596001276" width = 400> </a>

<h1 align=center><font size = 5>Clustering Construction Projects in Chicago</font></h1>

# Data Science Capstone Project

#### Author: Santiago Martínez Quintero
#### Civil Engineer
#### Contact: santiagomartinezqm@gmail.com
#### LinkedIn: https://www.linkedin.com/in/santiago-martinez-quintero/

___

## Briefing Words

Data Science is a field were all businesses are or are goin to be involved now and in the upcoming years and trying to use it in its favor to solve problems. Is an competitive advantage in order to get better results and to make your ideas happen. Is a tool that we cannot dismiss.

I started learning this field, because as a Civil Engineer, I have seen all the possibilities and oportunities that the construction sector has. It is really important if we want to succeed as a sector, be better and use all the tools that are helping the others. Of couse, as the *clichè* says, *Construction sector is years later than other sectors*. And I see it as an opportunity more than a problem. It is the right path.

Let's use it!

___

## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1. <a href="#item3">Introduction</a>

2. <a href="#item2">Data</a>

3. <a href="#item3">Modeling (Methodology)</a> 

4. <a href="#item4">Results</a> 

5. <a href="#item5">Discussion</a> 

6. <a href="#item5">Conclusion</a> 

7. <a href="#item5">References</a>   
</font>
</div>

## 1. Introduction

According to what we already said, Data Science is a very important field that is dveloped and is going to be in a good path in more sectors than the ones that now are already using this kind of analysis and decision making. Because of that, I am trying to improve construction and here is an example of what we could do, to have better decisions as construction managers.

In this Notebook, we will use the information of the Construction Projects that are being constructed in the City of Chicago and cluster them in order to forecast which ones are the best, because of the venues near them and predict which ones probably are going to have better sales and understand and identify, which zones are better to develop a housing construction project in that city.

### 1.1. Interests

Lots of project managers and builders, could and probably be interested in this kind of analysis because of it's value, not only for being ahead of the competece, but avoid reprocesses chosing the best locations and kind of projects that we are going to develope. And the analysis that we are going to see following, is just an small piece of what we can do.

### 1.2.  Data Science Methodology

It is important to note, that we are going to develope all the **Data Science Methdology** that Coursera and IBM taught in the professional certificate program in Data science, with have the following steps:

1. Bussiness Understanding. (Introduction)
2. Analytic Approach. (Interests).
3. Data:
    - Requirements
    - Collection
    - Understanding
    - Preparation
4. Modeling
5. Results
6. Discussion (Evaluation)
7. Conclussion (Feedback)
    
Also, this is better undertanded with this image given by IBM.



**(Note: most of the tables here in this Notebook, may look incomplete, but it is just that we printed an small part for better understanding)**

## 2. Data

### 2.1. Data Information

Now, according to the problem explaied before, we are going to get al the information of the construction projects that are being developed in Chicago, USA. As it is important and very well done, all the infromation is of free access to the people and citizens and you can find it in the following link: https://data.cityofchicago.org/Community-Economic-Development/Affordable-Rental-Housing-Developments-Map/k3g7-7kgc .

I am grateful that this information is free and I will use it for academic purposes in this Notebook as told before. 
First, we are going to examine the data to understand it. Second, we are going clean it if it is needed because of missing or wrong data. And finally, use the location of each project to put it in the Foursquare map, before starting with pre-processing the data.

The most important inormation that we are going to use, are the latitude and longitude coordinates, tha name and type of project and of course, the venues information that will come from Foursquare API.

### 2.2. Data Collection

What we are going to start doing is downloading the data from the souse given in order to understand it before any calculation. In order to do it, we have to import the libraries, that we imported the libraries that the model will need to develope the model.

In [1]:
#Import tha libraries that we are going to use
import itertools
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.ticker import NullFormatter
import pandas as pd
import numpy as np
from sklearn.cluster import KMeans 
import json # library to handle JSON files

In [2]:
!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1g             |       h516909a_1         2.1 MB  conda-forge
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    geopy-2.0.0                |     pyh9f0ad1d_0          63 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0          conda-forge
    geopy:           

In [8]:
! pip install folium

     |████████████████████████████████| 102kB 8.0MB/s ta 0:00:011


In [9]:
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


One part of the data is given below on a table.

In [24]:
# The code was removed by Watson Studio for sharing.

Community Area Name  Community Area Number Property Type  \
0             Douglas                     35           ARO   
1        Logan Square                     22   Multifamily   
2     Near North Side                      8           ARO   
3      Near West Side                     28           ARO   
4      West Englewood                     67   Multifamily   
5           Englewood                     68   Multifamily   
6              Uptown                      3        Senior   
7       Humboldt Park                     23   Multifamily   
8     Grand Boulevard                     38   Multifamily   
9      South Lawndale                     30   Multifamily   

                      Property Name                Address  Zip Code  \
0                     Lake Park LLC        243 E. 31st St.     60616   
1                Tierra Linda Apts.    1802 N. Sawyer Ave.     60647   
2                 Onni Atrium Devp.        202 W. Hill St.     60610   
3                        The Duncan     1515 W. Monroe St.     60607   
4          New West Englewood Homes       2109 W. 63rd St.     60636   
5                  Antioch Homes II  301 W. Marquette Road     60621   
6                Ruth Shriman House  4040 N. Sheridan Road     60613   
7                Tierra Linda Apts.  1649-51 N. Drake Ave.     60647   
8  Legends South - Hansberry Square      4034 S. State St.     60609   
9                    Casa Sor Juana     2700 S. Drake Ave.     60623   

   Phone Number                  Management Company  Units  X Coordinate  \
0  773-466-8332                        D Properties      1           NaN   
1  773-276-5338                               LUCHA      3           NaN   
2  312-585-6776                          Onni Group     43           NaN   
3  217-779-5697                       FLATS Leasing     52           NaN   
4  773-434-4929            Interfaith Housing Corp.     12           NaN   
5  773-994-4546  Universal Management Service, Inc.     69  1.175445e+06   
6  773-665-2210            Interfaith Housing Corp.    231  1.168833e+06   
7  773-276-5338                               LUCHA      6           NaN   
8  773-548-8800    Interstate Realty Management Co.    181  1.176908e+06   
9  312-850-1935            The Resurrection Project     29  1.153098e+06   

   Y Coordinate   Latitude  Longitude                         Location  
0           NaN        NaN        NaN                              NaN  
1           NaN        NaN        NaN                              NaN  
2           NaN        NaN        NaN                              NaN  
3           NaN        NaN        NaN                              NaN  
4           NaN        NaN        NaN                              NaN  
5  1.860492e+06  41.772564 -87.632419  (41.7725637689, -87.6324193182)  
6  1.927172e+06  41.955685 -87.654725  (41.9556850173, -87.6547251281)  
7           NaN        NaN        NaN                              NaN  
8  1.878123e+06  41.820911 -87.626526  (41.8209108342, -87.6265256539)  
9  1.885798e+06  41.842477 -87.713667    (41.842476558, -87.713667149)

We see that e have the location given by the are name, the direction and the latitude and longitude coordinates. Also, we have the information of the company who developed each project and the class of each project. With all these information we can cluster each project and then classified them and decide which ones are going to have better success and also which company is doing better accorss Chicago againts its competitors.

### 2.3. Data Cleaning
Here, before goin forward with the data, we have to cleant it because it can have missing data or errors. In this case, we can see that we don't have some location data of some projects. Because of that, we have to drop those ceells, because we cannot work with this data in Forsquare without latitude and longitude.

In [25]:
#Look the shape of the DataFrame Before drop Nan Rows
df_dataHousing.shape

(915, 14)

In [26]:
df_dataHousing.dropna(subset = ["Location"], inplace=True)
df_dataHousing.head()

Community Area Name  Community Area Number Property Type  \
5            Englewood                     68   Multifamily   
6               Uptown                      3        Senior   
8      Grand Boulevard                     38   Multifamily   
9       South Lawndale                     30   Multifamily   
10         Rogers Park                      1        Senior   

                       Property Name                Address  Zip Code  \
5                   Antioch Homes II  301 W. Marquette Road     60621   
6                 Ruth Shriman House  4040 N. Sheridan Road     60613   
8   Legends South - Hansberry Square      4034 S. State St.     60609   
9                     Casa Sor Juana     2700 S. Drake Ave.     60623   
10              Wayne Senior Housing     6928 N. Wayne Ave.     60626   

    Phone Number                  Management Company  Units  X Coordinate  \
5   773-994-4546  Universal Management Service, Inc.     69  1.175445e+06   
6   773-665-2210            Interfaith Housing Corp.    231  1.168833e+06   
8   773-548-8800    Interstate Realty Management Co.    181  1.176908e+06   
9   312-850-1935            The Resurrection Project     29  1.153098e+06   
10  773-274-0177                      CIG Management    119  1.165844e+06   

    Y Coordinate   Latitude  Longitude                         Location  
5   1.860492e+06  41.772564 -87.632419  (41.7725637689, -87.6324193182)  
6   1.927172e+06  41.955685 -87.654725  (41.9556850173, -87.6547251281)  
8   1.878123e+06  41.820911 -87.626526  (41.8209108342, -87.6265256539)  
9   1.885798e+06  41.842477 -87.713667    (41.842476558, -87.713667149)  
10  1.946059e+06  42.007574 -87.665171  (42.0075737709, -87.6651711448)

In [27]:
df_dataHousing.shape

(776, 14)

After dropping those cells, we noticed that the shape of teh data set changed. That is ok and doesn't change our purposes, we still have 776 projects to evaluate. This quantity es also going to be reviwed, because we cannot have duplocates and we have to drop if there is any.

In [46]:
df_dataHousing1 = df_dataHousing.drop_duplicates(subset=['Location'])
df_dataHousing1.head()

Community Area Name  Community Area Number Property Type  \
5            Englewood                     68   Multifamily   
6               Uptown                      3        Senior   
8      Grand Boulevard                     38   Multifamily   
9       South Lawndale                     30   Multifamily   
10         Rogers Park                      1        Senior   

                       Property Name                Address  Zip Code  \
5                   Antioch Homes II  301 W. Marquette Road     60621   
6                 Ruth Shriman House  4040 N. Sheridan Road     60613   
8   Legends South - Hansberry Square      4034 S. State St.     60609   
9                     Casa Sor Juana     2700 S. Drake Ave.     60623   
10              Wayne Senior Housing     6928 N. Wayne Ave.     60626   

    Phone Number                  Management Company  Units  X Coordinate  \
5   773-994-4546  Universal Management Service, Inc.     69  1.175445e+06   
6   773-665-2210            Interfaith Housing Corp.    231  1.168833e+06   
8   773-548-8800    Interstate Realty Management Co.    181  1.176908e+06   
9   312-850-1935            The Resurrection Project     29  1.153098e+06   
10  773-274-0177                      CIG Management    119  1.165844e+06   

    Y Coordinate   Latitude  Longitude                         Location  
5   1.860492e+06  41.772564 -87.632419  (41.7725637689, -87.6324193182)  
6   1.927172e+06  41.955685 -87.654725  (41.9556850173, -87.6547251281)  
8   1.878123e+06  41.820911 -87.626526  (41.8209108342, -87.6265256539)  
9   1.885798e+06  41.842477 -87.713667    (41.842476558, -87.713667149)  
10  1.946059e+06  42.007574 -87.665171  (42.0075737709, -87.6651711448)

In [47]:
df_dataHousing1.shape

(377, 14)

We dropped duplicate columns using the location, because it cant be for the same exact location. However, we are going to verify if there are more duplicates, counting by the project name.

In [48]:
df_dataHousing1['Property Name'].value_counts().head(20)

Park Douglas                       17
Lakefront Phase II                 13
Nelson Mandela Apts.               13
Rosa Parks Apts.                    8
Cornerstone Apts.                   5
65th Street Apts.                   5
Park Boulevard IIA                  4
Zapata Apts.                        4
St. Edmund's Oasis                  4
Park Boulevard IIB                  4
La Estancia                         3
Rebecca Walker Complex              3
Uptown Preservation Apts.           3
Resurrection Homes                  3
Hilliard Homes Phase I              2
Victory Center of South Chicago     2
Sunnyside Kenmore Apts.             2
Borinquen Bella Apts.               2
Parkside of Old Town IIB            2
Joy Plaza Residences                1
Name: Property Name, dtype: int64

We saw that for names, there are duplicates. So lets check more closely.

In [49]:
example=df_dataHousing1.ix[df_dataHousing1['Property Name']=='Park Douglas']
example.head()

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  if __name__ == '__main__':


Community Area Name  Community Area Number Property Type Property Name  \
20       North Lawndale                     29   Multifamily  Park Douglas   
29       North Lawndale                     29   Multifamily  Park Douglas   
67       North Lawndale                     29   Multifamily  Park Douglas   
72       North Lawndale                     29   Multifamily  Park Douglas   
114      North Lawndale                     29   Multifamily  Park Douglas   

                       Address  Zip Code  Phone Number  \
20   1313-15 S. Fairfiled Ave.     60608  773-893-4800   
29   1256-60 S. Washtenaw Ave.     60608  773-893-4800   
67         2642-44 W. 12 th Pl     60608  773-893-4800   
72      1224 S. Washtenaw Ave.     60608  773-893-4800   
114        1220 S. Talman Ave.     60608  773-893-4800   

                   Management Company  Units  X Coordinate  Y Coordinate  \
20   Interstate Realty Management Co.      4  1.158282e+06  1.893762e+06   
29   Interstate Realty Management Co.      1  1.158530e+06  1.893988e+06   
67   Interstate Realty Management Co.     17  1.158737e+06  1.894328e+06   
72   Interstate Realty Management Co.     31  1.158522e+06  1.894294e+06   
114  Interstate Realty Management Co.     10  1.158852e+06  1.894355e+06   

      Latitude  Longitude                         Location  
20   41.864228 -87.694428   (41.8642284474, -87.694427661)  
29   41.864843 -87.693509  (41.8648427949, -87.6935089238)  
67   41.865771 -87.692741  (41.8657711435, -87.6927414311)  
72   41.865682 -87.693533  (41.8656817278, -87.6935330428)  
114  41.865844 -87.692320   (41.865843892, -87.6923196475)

We saw, that eventhough there are lots of Projects with the same name, are different; but in essence is one massive made by the same company, for that reason, for our purposes in this general analysis, we are goin to drop duplicates and use only one which can give a general answe for all of them.

In [50]:
df_dataHousing1 = df_dataHousing.drop_duplicates(subset=['Property Name'])
df_dataHousing1.head()

Community Area Name  Community Area Number Property Type  \
5            Englewood                     68   Multifamily   
6               Uptown                      3        Senior   
8      Grand Boulevard                     38   Multifamily   
9       South Lawndale                     30   Multifamily   
10         Rogers Park                      1        Senior   

                       Property Name                Address  Zip Code  \
5                   Antioch Homes II  301 W. Marquette Road     60621   
6                 Ruth Shriman House  4040 N. Sheridan Road     60613   
8   Legends South - Hansberry Square      4034 S. State St.     60609   
9                     Casa Sor Juana     2700 S. Drake Ave.     60623   
10              Wayne Senior Housing     6928 N. Wayne Ave.     60626   

    Phone Number                  Management Company  Units  X Coordinate  \
5   773-994-4546  Universal Management Service, Inc.     69  1.175445e+06   
6   773-665-2210            Interfaith Housing Corp.    231  1.168833e+06   
8   773-548-8800    Interstate Realty Management Co.    181  1.176908e+06   
9   312-850-1935            The Resurrection Project     29  1.153098e+06   
10  773-274-0177                      CIG Management    119  1.165844e+06   

    Y Coordinate   Latitude  Longitude                         Location  
5   1.860492e+06  41.772564 -87.632419  (41.7725637689, -87.6324193182)  
6   1.927172e+06  41.955685 -87.654725  (41.9556850173, -87.6547251281)  
8   1.878123e+06  41.820911 -87.626526  (41.8209108342, -87.6265256539)  
9   1.885798e+06  41.842477 -87.713667    (41.842476558, -87.713667149)  
10  1.946059e+06  42.007574 -87.665171  (42.0075737709, -87.6651711448)

In [51]:
df_dataHousing1.shape

(305, 14)

In [52]:
df_dataHousing1 = df_dataHousing1.drop_duplicates(subset=['Location'])
df_dataHousing1.head()

Community Area Name  Community Area Number Property Type  \
5            Englewood                     68   Multifamily   
6               Uptown                      3        Senior   
8      Grand Boulevard                     38   Multifamily   
9       South Lawndale                     30   Multifamily   
10         Rogers Park                      1        Senior   

                       Property Name                Address  Zip Code  \
5                   Antioch Homes II  301 W. Marquette Road     60621   
6                 Ruth Shriman House  4040 N. Sheridan Road     60613   
8   Legends South - Hansberry Square      4034 S. State St.     60609   
9                     Casa Sor Juana     2700 S. Drake Ave.     60623   
10              Wayne Senior Housing     6928 N. Wayne Ave.     60626   

    Phone Number                  Management Company  Units  X Coordinate  \
5   773-994-4546  Universal Management Service, Inc.     69  1.175445e+06   
6   773-665-2210            Interfaith Housing Corp.    231  1.168833e+06   
8   773-548-8800    Interstate Realty Management Co.    181  1.176908e+06   
9   312-850-1935            The Resurrection Project     29  1.153098e+06   
10  773-274-0177                      CIG Management    119  1.165844e+06   

    Y Coordinate   Latitude  Longitude                         Location  
5   1.860492e+06  41.772564 -87.632419  (41.7725637689, -87.6324193182)  
6   1.927172e+06  41.955685 -87.654725  (41.9556850173, -87.6547251281)  
8   1.878123e+06  41.820911 -87.626526  (41.8209108342, -87.6265256539)  
9   1.885798e+06  41.842477 -87.713667    (41.842476558, -87.713667149)  
10  1.946059e+06  42.007574 -87.665171  (42.0075737709, -87.6651711448)

In [53]:
df_dataHousing1.shape

(296, 14)

After cleaning the dataset, we sow that the size of the of the dataset is just a part of the first one. But in this case, is perfect for our purposes and better, because we are going to get better answers without data. For the clusters maybe some duplicates were not a problem, but for taking a step from there and distribute a select the better sector and companies, yes. Taking that into account is really important for this analysis to understand that we cannot have issues on the dataset and we also have to check until we are ok with it.

## 3. Modeling (Methodology)

### 3.1. Understanding the location data. (Data Preparation)
In order to start modeling, we have first to understand how is distributed the location data and check if there is no problem with the Foursquare API and also with the points in the dataset that we claned before. This part is very important in Data Science and for this project, because we have to check if our most important input is working ok and is in the location that we expected.

In [60]:
#Get the Chicago Coordinates
address = 'Chicago, IL'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude1 = location.latitude
longitude1 = location.longitude
print('The geograpical coordinate of Chincago are {}, {}.'.format(latitude1, longitude1))

The geograpical coordinate of Chincago are 41.8755616, -87.6244212.


Afer running the geolocator package in python, we got the latitude and longitude of the city of chicago, without seraching ir manually. With this, we are going to make the first example, where we are goint to print the map with all the hounsing projects in it and take a general view of what we are going to face. You can se that following:

In [64]:
# create map of Chicago using latitude and longitude values
map_chicago = folium.Map(location=[latitude1, longitude1], zoom_start=10)

# add markers to map
for lat, lng, areaname, propname in zip(df_dataHousing1['Latitude'], df_dataHousing1['Longitude'], df_dataHousing1['Community Area Name'], df_dataHousing1['Property Name']):
    label = '{}, {}'.format(areaname, propname)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='red',
        fill=True,
        fill_color='#FF5733',
        fill_opacity=0.7,
        parse_html=False).add_to(map_chicago)  
    
map_chicago

We saw that we got a large number of projects, well distributed around the city. For that reason we think we are ging to get more interesting answers in the end. Also, we saw that no project is in the same exact location of another one.

Also, before we model per se, we are going to get all the venues using the Foursquare API, which help us, location all the vanues with an specific radius and a limit of venues, given the top ones.

For that reason we are going to asume, 100 meters of radius and the 20 top venues of each location data. We don't want to get more, just for computing costs.

In [65]:
#Define credentials of Foursquare API
CLIENT_ID = 'R4GLZICD2IXI4Q0N1DG4T2GZ1ZMJKBCBGBRVM4J2BDKBHPMR' # your Foursquare ID
CLIENT_SECRET = 'NUJA3UJ4FIQLTAFJOOVCDEGMGEWNR3WEXWI5G3OINOHQWMAD' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: R4GLZICD2IXI4Q0N1DG4T2GZ1ZMJKBCBGBRVM4J2BDKBHPMR
CLIENT_SECRET:NUJA3UJ4FIQLTAFJOOVCDEGMGEWNR3WEXWI5G3OINOHQWMAD


In [66]:
LIMIT = 20 # limit of number of venues returned by Foursquare API
radius = 100 # define radius
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Property Name', 
                  'Property Name Latitude', 
                  'Property Name Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [67]:
#Create a Dataframe using the vanues
chicago_venues = getNearbyVenues(names=df_dataHousing1['Property Name'],
                                   latitudes=df_dataHousing1['Latitude'],
                                   longitudes=df_dataHousing1['Longitude']
                                  )

Antioch Homes II
Ruth Shriman House
Legends South - Hansberry Square
Casa Sor Juana
Wayne Senior Housing
Hilliard Homes Phase II
Mayfair Commons
Indiana Manor Town Homes
Senior Suites of South Shore
5822 North Western, LLC
The Westner
Sanctuary Place
Park Douglas
Lake Park Crescent
Senior Suites of Auburn-Gresham
65th Street Apts.
LUCHA Affordable Units
Roosevelt Square Phase I /ABLA Rental
St. Edmund's Corners
Butler Apts.
Resurrection Homes
Pine Central
Harriett Tubman Apts.
St. Edmund's Village
East Park Apts.
Rosa Parks Apts.
Victory Center of South Chicago
Zapata Apts.
Kennedy Jordan Manor
Park Boulevard IIB
2917-39 N. Central
Roosevelt Towers
Dicksons Estates Apts.
Oakwood Shores Terrace
Park Boulevard IB
Senior Suites of Belmont-Cragin
Coppin House
Homan Square Apartments Phase IV
Congress Commons LLC
1955 Richmond Place
Uptown Preservation Apts.
Senior Suites of Hegewisch
West Humboldt Place
Porta Coeli Residence
4801 N. Ravenswood Ave.
Viridian on Sheridan
Noca Blu
Rebecca Wal

In [68]:
print(chicago_venues.shape)
chicago_venues.head()

(3973, 7)


Property Name  Property Name Latitude  Property Name Longitude  \
0    Antioch Homes II               41.772564               -87.632419   
1    Antioch Homes II               41.772564               -87.632419   
2    Antioch Homes II               41.772564               -87.632419   
3  Ruth Shriman House               41.955685               -87.654725   
4  Ruth Shriman House               41.955685               -87.654725   

                  Venue  Venue Latitude  Venue Longitude      Venue Category  
0     Fishermans Island       41.772994       -87.629602  Seafood Restaurant  
1                 Shell       41.772891       -87.629576         Gas Station  
2       Periwinkle Park       41.774640       -87.628187                Park  
3  Byron's Hot Dog Haus       41.954397       -87.655252       Hot Dog Joint  
4     PR Italian Bistro       41.953050       -87.654541  Italian Restaurant

Now we got the information of all the venues for each Project (*Property Name*), we are going to group them with 3 steps.

First, we are goint to use **One Hot Encoding**, to get a unique value true, for each row, according to all the venues that we got. And, What is One Hot Encoding? 

Accordingto Sarah and David Harris, in the chapter of *One Hot Encoding* of their Book *Digital Design and Computer Architecture*, One Hot Encoding is:"*The “one hot” encoding scheme refers to the fact that each state in a state machine has its own state variable in the form of a flip-flop, and only one state variable may be active (“hot”) at any particular time.
View chapterPurchase book*". (Took the 1 of August of 2020, from: https://www.sciencedirect.com/topics/computer-science/one-hot-encoding).

You can see it as follows:

In [87]:
# one hot encoding
chicago_onehot = pd.get_dummies(chicago_venues[['Venue Category']], prefix="", prefix_sep="")
chicago_onehot.head()

ATM  Accessories Store  African Restaurant  American Restaurant  \
0    0                  0                   0                    0   
1    0                  0                   0                    0   
2    0                  0                   0                    0   
3    0                  0                   0                    0   
4    0                  0                   0                    0   

   Antique Shop  Arcade  Argentinian Restaurant  Art Gallery  Art Museum  \
0             0       0                       0            0           0   
1             0       0                       0            0           0   
2             0       0                       0            0           0   
3             0       0                       0            0           0   
4             0       0                       0            0           0   

   Arts & Crafts Store  ...  Warehouse  Warehouse Store  Weight Loss Center  \
0                    0  ...          0                0                   0   
1                    0  ...          0                0                   0   
2                    0  ...          0                0                   0   
3                    0  ...          0                0                   0   
4                    0  ...          0                0                   0   

   Whisky Bar  Wine Bar  Wine Shop  Winery  Wings Joint  Women's Store  \
0           0         0          0       0            0              0   
1           0         0          0       0            0              0   
2           0         0          0       0            0              0   
3           0         0          0       0            0              0   
4           0         0          0       0            0              0   

   Yoga Studio  
0            0  
1            0  
2            0  
3            0  
4            0  

[5 rows x 294 columns]

Here, you can see how is dristruted by dummies (1's and 0's), showing that each row has an unique value of a venue (1). (True = 1, False = 0).

The second step: it is better if we see the Property Name in the first column of each row, as follows:

In [88]:
# add property name column back to dataframe
chicago_onehot['Property Name'] = chicago_venues['Property Name'] 
chicago_onehot.head()

ATM  Accessories Store  African Restaurant  American Restaurant  \
0    0                  0                   0                    0   
1    0                  0                   0                    0   
2    0                  0                   0                    0   
3    0                  0                   0                    0   
4    0                  0                   0                    0   

   Antique Shop  Arcade  Argentinian Restaurant  Art Gallery  Art Museum  \
0             0       0                       0            0           0   
1             0       0                       0            0           0   
2             0       0                       0            0           0   
3             0       0                       0            0           0   
4             0       0                       0            0           0   

   Arts & Crafts Store  ...  Warehouse Store  Weight Loss Center  Whisky Bar  \
0                    0  ...                0                   0           0   
1                    0  ...                0                   0           0   
2                    0  ...                0                   0           0   
3                    0  ...                0                   0           0   
4                    0  ...                0                   0           0   

   Wine Bar  Wine Shop  Winery  Wings Joint  Women's Store  Yoga Studio  \
0         0          0       0            0              0            0   
1         0          0       0            0              0            0   
2         0          0       0            0              0            0   
3         0          0       0            0              0            0   
4         0          0       0            0              0            0   

        Property Name  
0    Antioch Homes II  
1    Antioch Homes II  
2    Antioch Homes II  
3  Ruth Shriman House  
4  Ruth Shriman House  

[5 rows x 295 columns]

In [89]:
# move property name column to the first column
fixed_columns = [chicago_onehot.columns[-1]] + list(chicago_onehot.columns[:-1])
chicago_onehot = chicago_onehot[fixed_columns]
chicago_onehot.head()

Property Name  ATM  Accessories Store  African Restaurant  \
0    Antioch Homes II    0                  0                   0   
1    Antioch Homes II    0                  0                   0   
2    Antioch Homes II    0                  0                   0   
3  Ruth Shriman House    0                  0                   0   
4  Ruth Shriman House    0                  0                   0   

   American Restaurant  Antique Shop  Arcade  Argentinian Restaurant  \
0                    0             0       0                       0   
1                    0             0       0                       0   
2                    0             0       0                       0   
3                    0             0       0                       0   
4                    0             0       0                       0   

   Art Gallery  Art Museum  ...  Warehouse  Warehouse Store  \
0            0           0  ...          0                0   
1            0           0  ...          0                0   
2            0           0  ...          0                0   
3            0           0  ...          0                0   
4            0           0  ...          0                0   

   Weight Loss Center  Whisky Bar  Wine Bar  Wine Shop  Winery  Wings Joint  \
0                   0           0         0          0       0            0   
1                   0           0         0          0       0            0   
2                   0           0         0          0       0            0   
3                   0           0         0          0       0            0   
4                   0           0         0          0       0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 295 columns]

Much better, Don't you think?

Third, we have to group all the information by *Property Name*, in order to get unique values. In this case, we are going to gruoup them by this column and using the mean of each one of the others, in order to get the most common.

In [93]:
#Group all the information by Property Name and mean.
chicago_grouped = chicago_onehot.groupby('Property Name').mean().reset_index()
chicago_grouped.head(10)

Property Name  ATM  Accessories Store  \
0                  1038 N. Ashland  0.0                0.0   
1  1237 N. Milwaukee Ave.Owner LLC  0.0                0.0   
2               1611 West Division  0.0                0.0   
3       1911-21 W. Irving Park Rd.  0.0                0.0   
4              1955 Richmond Place  0.0                0.0   
5           2211 N. Milwaukee Ave.  0.0                0.0   
6                    2400 Cortland  0.0                0.0   
7                    2501 Armitage  0.0                0.0   
8               2917-39 N. Central  0.0                0.0   
9            2931-57 N. Harlem LLC  0.0                0.0   

   African Restaurant  American Restaurant  Antique Shop  Arcade  \
0                 0.0                 0.00           0.0     0.0   
1                 0.0                 0.00           0.0     0.0   
2                 0.0                 0.00           0.0     0.0   
3                 0.0                 0.00           0.0     0.0   
4                 0.0                 0.00           0.0     0.0   
5                 0.0                 0.00           0.0     0.0   
6                 0.0                 0.00           0.0     0.0   
7                 0.0                 0.00           0.0     0.0   
8                 0.0                 0.05           0.0     0.0   
9                 0.0                 0.00           0.0     0.0   

   Argentinian Restaurant  Art Gallery  Art Museum  ...  Warehouse  \
0                    0.05          0.0         0.0  ...        0.0   
1                    0.05          0.0         0.0  ...        0.0   
2                    0.05          0.0         0.0  ...        0.0   
3                    0.00          0.0         0.0  ...        0.0   
4                    0.00          0.0         0.0  ...        0.0   
5                    0.00          0.0         0.0  ...        0.0   
6                    0.00          0.0         0.0  ...        0.0   
7                    0.00          0.0         0.0  ...        0.0   
8                    0.00          0.0         0.0  ...        0.0   
9                    0.00          0.0         0.0  ...        0.0   

   Warehouse Store  Weight Loss Center  Whisky Bar  Wine Bar  Wine Shop  \
0              0.0                 0.0         0.0       0.0       0.00   
1              0.0                 0.0         0.0       0.0       0.00   
2              0.0                 0.0         0.0       0.0       0.00   
3              0.0                 0.0         0.0       0.0       0.00   
4              0.0                 0.0         0.0       0.0       0.00   
5              0.0                 0.0         0.0       0.0       0.00   
6              0.0                 0.0         0.0       0.0       0.05   
7              0.0                 0.0         0.0       0.0       0.00   
8              0.0                 0.0         0.0       0.0       0.00   
9              0.0                 0.0         0.0       0.0       0.00   

   Winery  Wings Joint  Women's Store  Yoga Studio  
0     0.0          0.0            0.0         0.00  
1     0.0          0.0            0.0         0.00  
2     0.0          0.0            0.0         0.00  
3     0.0          0.0            0.0         0.05  
4     0.0          0.0            0.0         0.00  
5     0.0          0.0            0.0         0.05  
6     0.0          0.0            0.0         0.05  
7     0.0          0.0            0.0         0.05  
8     0.0          0.0            0.0         0.00  
9     0.0          0.0            0.0         0.00  

[10 rows x 295 columns]

Finally, we got all the information that we needed. Now, we can get some modeling Hypothesis, before starting doing it.

### 3.2. Modeling Hypothesis

Now, we have seen the dat distributed in the map, we can star making some hypothesis of the answers that we are going to get, for example:

1. The clusters are going to be distributed clearly by different sector of the city.
2. The senior projects, are going to be near the lake, because of the tranquility of the home owners which are going to be old people.
3. The multifamiliar properties, are going to be near the center of the city, in order to be close enought to the important venues, as banks, parks, restaurants, malls, etc.
4. The build companies, are going to be disperced around the city, without showing any particular sector of interest.
5. The large projects (number of units), are going to be located near each other.

So from now, lest start modeling in order to check all these hypothesis.

### 3.3. Modeling

For modeling, we are going to use k-Means mathod, which is an unsupervised method for clustering. This mathos is called unsupervised, because the method makes inferences from vectors or more simple, from the information given in the dataset automatically. The process, is given by an iterative process where  the data is clustered around given centroids, which with the iterative calculate, move its position, to get the better result and minimize the error. (Garbade, M)(Found the 1 of August of 2020 from (09/2018) *Understanding K-means Clustering in Machine Learning*  from: https://towardsdatascience.com/understanding-k-means-clustering-in-machine-learning-6a6e67336aa1).

Given the theory and the data ready before, we are going to select the top 10 venues, from those 20 that we have in the table for each *Property Name*. Also we are going to set a second input, which is a number of clusters of 7.

- Inputs:

    - Number of Top Venues = 10
    - Clusters = 7
    
The Output is a table that sorts the Property Names and its top 10 venues

In [94]:
# Funtion that sorts the top venues of the input
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [96]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Property Name']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
property_venues_sorted = pd.DataFrame(columns=columns)
property_venues_sorted['Property Name'] = chicago_grouped['Property Name']

for ind in np.arange(chicago_grouped.shape[0]):
    property_venues_sorted.iloc[ind, 1:] = return_most_common_venues(chicago_grouped.iloc[ind, :], num_top_venues)

property_venues_sorted.head(10)

Property Name 1st Most Common Venue  \
0                  1038 N. Ashland    Mexican Restaurant   
1  1237 N. Milwaukee Ave.Owner LLC               Theater   
2               1611 West Division    Mexican Restaurant   
3       1911-21 W. Irving Park Rd.                   Bar   
4              1955 Richmond Place    Mexican Restaurant   
5           2211 N. Milwaukee Ave.    Mexican Restaurant   
6                    2400 Cortland           Coffee Shop   
7                    2501 Armitage    Mexican Restaurant   
8               2917-39 N. Central    Mexican Restaurant   
9            2931-57 N. Harlem LLC  Fast Food Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Bar                Hostel                   Gym   
1        Cosmetics Shop  Gym / Fitness Center             Speakeasy   
2        Cosmetics Shop                   Gym             Speakeasy   
3               Brewery    Chinese Restaurant            Beer Store   
4           Pizza Place        Ice Cream Shop             Bike Shop   
5                   Bar    Salon / Barbershop      Sushi Restaurant   
6                   Bar                 Trail           Yoga Studio   
7         Hot Dog Joint           Coffee Shop           Yoga Studio   
8            Donut Shop        Discount Store           Gas Station   
9                   Bar              Pharmacy        Sandwich Place   

  5th Most Common Venue      6th Most Common Venue     7th Most Common Venue  \
0             Speakeasy                Snack Place          Sushi Restaurant   
1           Snack Place    New American Restaurant        Mexican Restaurant   
2           Snack Place           Sushi Restaurant        Seafood Restaurant   
3      Sushi Restaurant                   Boutique             Bowling Alley   
4              Pie Shop          Convenience Store              Dessert Shop   
5          Cocktail Bar                       Café                       Gym   
6      Asian Restaurant           Cuban Restaurant            Ice Cream Shop   
7            Comic Shop  Latin American Restaurant                     Trail   
8    Seafood Restaurant     Thrift / Vintage Store  Mediterranean Restaurant   
9       Automotive Shop         Mexican Restaurant               Video Store   

  8th Most Common Venue        9th Most Common Venue  \
0    Seafood Restaurant                 Optical Shop   
1      Sushi Restaurant           Falafel Restaurant   
2          Optical Shop  Eastern European Restaurant   
3                  Café                  Yoga Studio   
4          Cocktail Bar               Shipping Store   
5             Bike Shop                   Street Art   
6            Comic Shop                 Cocktail Bar   
7                   Bar             Sushi Restaurant   
8          Burger Joint               Sandwich Place   
9         Jewelry Store                 Liquor Store   

        10th Most Common Venue  
0  Eastern European Restaurant  
1  Eastern European Restaurant  
2           Salon / Barbershop  
3           Turkish Restaurant  
4                 Camera Store  
5      New American Restaurant  
6                   Board Shop  
7             Asian Restaurant  
8                   Laundromat  
9           Italian Restaurant

**Finally**, having that table (dataset), we are going to run the k-Means package in Python, which shows a vector of the clusters, which we are going to add to a table, in order to map the clusters (points), in the Chicago's map.

In [97]:
# set number of clusters
kclusters = 7

chicago_grouped_clustering = chicago_grouped.drop('Property Name', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(chicago_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 0, 1, 1, 1, 0, 3], dtype=int32)

## 4. Results

After having that vector, we have the answers of the Clustering. But we cannot read it easily, so we are going to create a table, with merge all teh information from the first dataset and the information of the 10 most common venues, and then map them.

In [98]:
# add clustering labels
property_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

chicago_merged = df_dataHousing1

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
chicago_merged = chicago_merged.join(property_venues_sorted.set_index('Property Name'), on='Property Name')

chicago_merged.head()

Community Area Name  Community Area Number Property Type  \
5            Englewood                     68   Multifamily   
6               Uptown                      3        Senior   
8      Grand Boulevard                     38   Multifamily   
9       South Lawndale                     30   Multifamily   
10         Rogers Park                      1        Senior   

                       Property Name                Address  Zip Code  \
5                   Antioch Homes II  301 W. Marquette Road     60621   
6                 Ruth Shriman House  4040 N. Sheridan Road     60613   
8   Legends South - Hansberry Square      4034 S. State St.     60609   
9                     Casa Sor Juana     2700 S. Drake Ave.     60623   
10              Wayne Senior Housing     6928 N. Wayne Ave.     60626   

    Phone Number                  Management Company  Units  X Coordinate  \
5   773-994-4546  Universal Management Service, Inc.     69  1.175445e+06   
6   773-665-2210            Interfaith Housing Corp.    231  1.168833e+06   
8   773-548-8800    Interstate Realty Management Co.    181  1.176908e+06   
9   312-850-1935            The Resurrection Project     29  1.153098e+06   
10  773-274-0177                      CIG Management    119  1.165844e+06   

    ...  1st Most Common Venue  2nd Most Common Venue  3rd Most Common Venue  \
5   ...                   Park            Gas Station     Seafood Restaurant   
6   ...            Coffee Shop          Hot Dog Joint                    Bar   
8   ...        Harbor / Marina      Convenience Store           Intersection   
9   ...     Mexican Restaurant      Mobile Phone Shop         Ice Cream Shop   
10  ...            Pizza Place                    Bar       Tapas Restaurant   

   4th Most Common Venue  5th Most Common Venue  \
5   Ethiopian Restaurant            Event Space   
6                 Bakery                   Park   
8     Italian Restaurant          Train Station   
9            Pizza Place                   Bank   
10        Farmers Market          Grocery Store   

              6th Most Common Venue        7th Most Common Venue  \
5                        Eye Doctor           Falafel Restaurant   
6                     Tattoo Parlor           Italian Restaurant   
8   Southern / Soul Food Restaurant                        Motel   
9                       Art Gallery                 Liquor Store   
10                      Coffee Shop  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
5                   Farm        Farmers Market            Yoga Studio  
6                Theater           Pizza Place      Indian Restaurant  
8     Light Rail Station           Yoga Studio             Eye Doctor  
9           Dessert Shop     Food & Drink Shop      Fondue Restaurant  
10                  Café      Asian Restaurant              Bookstore  

[5 rows x 25 columns]

And the maps of the last table, is following:

In [100]:
# create map
map_clusters = folium.Map(location=[latitude1, longitude1], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(chicago_merged['Latitude'], chicago_merged['Longitude'], chicago_merged['Property Name'], chicago_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

We got our clusters here.

Now, taking those clusters, we are going to Classify the clusters as follows: